feat(select,input,form-field): move test harnesses out of experimental

crisbeto opened the pull request:
Moves the test harnesses for the `select`, `input` and `form-field` packages out of `material-experimental`.

crisbeto commented on src/material/select/testing/select-harness.ts:
```diff
@@ -7,9 +7,7 @@
-} from '@angular/material-experimental/form-field/testing/control';
+import {MatFormFieldControlHarness} from '@angular/material/form-field/testing/control';
 import {SelectHarnessFilters} from './select-harness-filters';
 import {MatSelectOptionHarness, MatSelectOptionGroupHarness} from './option-harness';
```
Note that we have a harness for an option inside `mat-select`, as well as for an option in `mat-autocomplete`. Once this PR is merged in I can remove the duplication and add single, fully-featured harness for `mat-option` in `core`.

mmalerba replied:
It would be really nice if we could get that in before final. I forgot that autocomplete was already moved out of experimental with the split mat-option harness

crisbeto commented:
Setting to P2 since there's other work that depends on these changes.

mmalerba commented:
Sorry, it won't let me add comments inline since the lines are unedited, but here's a breakdown by class of changes I think we should make before moving this API out of experimental:

```
FormFieldHarnessFilters
	+ floatingLabelText?: string | RegExp
	+ hasErrors?: boolean

MatFormFieldHarness
	+ hasErrors(): Promise<boolean>
	+ getHarnessLoaderForInfix(): Promise<HarnessLoader>
	getPrefixContainer => getHarnessLoaderForPrefix
	getSuffixContainer => getHarnessLoaderForSuffix
	getErrorMessages => getTextErrors
	getHintMessages => getTextHints

InputHarnessFilters
	+ placeholder?: string | RegExp

SelectHarnessFilters
	+ placeholder?: string | RegExp
	+ value?: string | RegExp

MatSelectHarness
	+ select(filters?: SelectOptionHarnessFilters): Promise<void>
	getOptions => add optional filters
	getOptionGroups => add optional filters
```

Also, do a pass on the API docs, make sure they are consistent with other harnesses (in particular I noticed `with` method, move description of filter options to the filter class).

crisbeto commented:
@mmalerba I've updated the harnesses based on your feedback, aside from a couple of issues:

1. I'm not sure how `getHarnessLoaderForInfix(): Promise<HarnessLoader>` is supposed to work and what it would be useful for. We already have harnesses for the input and select which expose the content inside the infix.
2. I don't think we can do the `value` and `placeholder` filters for the select harness, because the values aren't always in the DOM so we can't extract them reliably.

mmalerba commented:
Discussed with @crisbeto on slack. After thinking about it more, I agree that those two don't make sense

mmalerba commented on src/material/select/testing/select-harness.ts:
```diff
@@ -107,6 +109,13 @@ export class MatSelectHarness extends MatFormFieldControlHarness {
+  /** Selects the options that match the passed-in filter. */
+  async select(filter: OptionHarnessFilters = {}): Promise<void> {
+    await this.open();
+    const options = await this.getOptions(filter);
```
should we check if its multiple selection and if not just click the first one?

crisbeto replied:
Good point. Changed.

crisbeto commented:
I added `getPanel`, because I thought that it might useful if somebody wants to get info that we haven't exposed directly in the harness. I've removed it, can you take a look @mmalerba?

crisbeto commented on src/material/select/testing/select-harness.ts:
```diff
@@ -107,6 +99,19 @@ export class MatSelectHarness extends MatFormFieldControlHarness {
+    const [isMultiple, options] = await Promise.all([this.isMultiple(), this.getOptions(filter)]);
+
+    if (isMultiple) {
+      await Promise.all(options.map(option => option.click()));
```
These are the custom `mat-option` elements so we don't have access to the select/deselect methods. Should I rename it to `click` instead?

mmalerba replied:
You can leave it as is for now and leave a TODO to split out separate select/deselect methods later

crisbeto commented:
@mmalerba I've addressed the latest set the feedback, aside from exposing the option harnesses. I haven't exposed or documented them properly yet, because their code is duplicated between `mat-select` and `mat-autocomplete`. Once this PR is merged in I'll add proper, fully-featured harnesses that are exposed through `material/core/testing`.

mmalerba commented:
I'm hesitant about just merging this and doing the option stuff in a follow up because I don't want this temporary state API to wind up in an RC, or worse final if the follow-up doesn't make it in in time. I would suggest doing the common option _first_ and putting this one on hold until that's in.

cc @jelbourn

crisbeto commented:
I can put together the option PR for Monday so we have some time to merge it in for final. I would've made a shared option harness to begin with, but the autocomplete and select harnesses started off in experimental and we don't have a `material-experimental/core` to put it in.

jelbourn commented:
I'm fine with doing this in two steps since I just did an RC today and will know not to do another one until we consolidate

mmalerba commented on src/material/form-field/testing/form-field-harness-filters.ts:
```diff
@@ -8,4 +8,7 @@
 import {BaseHarnessFilters} from '@angular/cdk/testing';
 
-export interface FormFieldHarnessFilters extends BaseHarnessFilters {}
+export interface FormFieldHarnessFilters extends BaseHarnessFilters {
```
Add doc comments to this

mmalerba commented on src/material/input/testing/input-harness-filters.ts:
```diff
@@ -9,5 +9,6 @@
 
 export interface InputHarnessFilters extends BaseHarnessFilters {
-  value?: string;
+  value?: string | RegExp;
```
Add doc comments

crisbeto commented:
Addressed the last few things @mmalerba.
